File: leetcode/0002. 两数相加/main.go
package main

type ListNode struct {
	Val  int
	Next *ListNode
}

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	var carry bool
	var sum int
	tail := &ListNode{}
	var result *ListNode

	for l1 != nil || l2 != nil {
		if carry {
			sum += 1
			carry = false
		}
		if l1 != nil {
			sum += l1.Val
			l1 = l1.Next
		}
		if l2 != nil {
			sum += l2.Val
			l2 = l2.Next
		}
		if sum > 9 {
			sum = sum % 10
			carry = true
		}
		if result == nil {
			result = &ListNode{Val: sum}
			tail = result
		} else {
			tail.Next = &ListNode{Val: sum}
			tail = tail.Next
		}
		sum = 0
	}
	if carry {
		tail.Next = &ListNode{Val: 1}
	}
	return result
}

func main() {
	addTwoNumbers(nil, nil)
}

File: leetcode/0002. 两数相加/main.js
function ListNode(val, next) {
    this.val = (val === undefined ? 0 : val)
    this.next = (next === undefined ? null : next)
}

var addTwoNumbers = function (l1, l2) {
    let carry = 0;
    let head = new ListNode();
    let curr = head;
    while (l1 || l2) {
        let sum = (l1 ? l1.val : 0) + (l2 ? l2.val : 0) + carry;
        carry = Math.floor(sum / 10);
        curr.next = new ListNode(sum % 10);
        curr = curr.next;
        if (l1) l1 = l1.next;
        if (l2) l2 = l2.next;
    }
    if (carry) curr.next = new ListNode(carry);
    return head.next;
};

(function main() {
    let l1 = new ListNode(2, ListNode(4, ListNode(3)));
    let l2 = new ListNode(5, ListNode(6, ListNode(4)));
    let l3 = addTwoNumbers(l1, l2);
    console.log(l3.val);
})()
